packages/react/src/components/penumbra-provider.tsx

```diff
@@ -0,0 +1,204 @@
+        }
+      }
+    };
+
+    const ac = new AbortController();
+    providerInjection?.addEventListener('penumbrastate', listener, {
+      signal: ac.signal,
+    });
+    return () => ac.abort();
+  }, [providerInjection, providerInjection?.addEventListener, manifest, failure]);
+
+  // fetch manifest to confirm presence of provider
+  useEffect(() => {
+    // require provider
+    if (!providerOrigin || !providerInjection) {
+      return;
+    }
+    // don't repeat, unnecessary if failed
+    if (!!manifest || failure) {
+      return;
+    }
+
+    // sync assertion
+    try {
+      assertManifestOrigin(providerOrigin, providerInjection);
+    } catch (cause) {
+      setFailureUnknown(cause);
+      return;
+    }
+
+    // abortable fetch
+    const ac = new AbortController();
+    const fetchManifest = fetch(providerInjection.manifest, { signal: ac.signal }).catch(
+      (noAbortError: unknown) => {
+        // abort is not a failure
+        if (noAbortError instanceof Error && noAbortError.name === 'AbortError') {
+          return;
+        } else {
+          throw noAbortError;
+        }
+      },
+    );
+
+    // async handle response
+    void fetchManifest
+      .then(async res => {
+        const manifestJson: unknown = await res?.json();
+        if (manifestJson) {
+          // this cast is fairly safe coming from an extension manifest, where
+          // schema is enforced by chrome store.
+          setManifest(manifestJson as PenumbraManifest);
+        }
+      })
+      .catch(setFailureUnknown);
+
+    return () => ac.abort();
+  }, [providerOrigin, providerInjection, manifest, setManifest]);
+
+  // request effect
+  useEffect(() => {
+    // require manifest, no failures
+    if (!manifest || failure) {
+      return;
+    }
+
+    switch (providerState) {
+      case PenumbraInjectionState.Present:
+        if (makeApprovalRequest) {
+          void providerInjection?.request().catch(setFailureUnknown);
+        }
+        break;
+      default:
+        break;
+    }
+  }, [makeApprovalRequest, providerState, providerInjection?.request, manifest, failure]);
+
+  // connect effect
+  useEffect(() => {
+    // require manifest, no failures
+    if (!manifest || failure) {
+      return;
+    }
+
+    switch (providerState) {
+      case PenumbraInjectionState.Present:
+        if (!makeApprovalRequest) {
+          void providerInjection
+            ?.connect()
+            .then(p => setProviderPort(p))
+            .catch(setFailureUnknown);
+        }
+        break;
+      case PenumbraInjectionState.Requested:
+        void providerInjection
+          ?.connect()
+          .then(p => setProviderPort(p))
+          .catch(setFailureUnknown);
+        break;
+      default:
+        break;
+    }
+  }, [makeApprovalRequest, providerState, providerInjection?.connect, manifest, failure]);
+
+  const createdContext: PenumbraContext = useMemo(
```

Comment:
Do we need to have context? It would be nice if the consumer wouldn't need to wrap their app with a provider. Especially since it appears that this initiates actions automatically. Doesn't all state live within the injection now? Aka, can't we just rely upon a `usePenumbra()` hook without a provider?

Reply:
if the developer just wants to grab a client wherever they are, `@penumbra-zone/client` already provides methods for that.

```ts
import { createPenumbraClient } from '@penumbra-zone/client/create';
import { ViewService } from '@penumbra-zone/protobuf';

// this goes through the same connection process
const viewClient = createPenumbraClient(ViewService, someOptionalOrigin);

export const ChainIdComponent = () => {
    const [chainId, setChainId] = useState<string>();
    useEffect(() => { // or useQuery or whatever
        const { parameters } = await viewClient.appParameters({});
        return setChainId(parameters?.chainId);
    });
    return <div>{chainId}</div>
};
```

i agree that a dedicated react package isn't much of a value-add. at best this may help prevent a naive developer from constructing/deconstructing a transport and client for every component on every rerender. 

yes the injection tracks its own detailed connection state including called methods, so now the caller doesn't have to remember what they've done. but all that it delivers is a `MessagePort`. the page still needs to construct its own transport etc.

and if the goal is to share state among components, context providers are the way to achieve shared state in react, unless we expect users to prop-drill the clients everywhere. and it's convenient to throw in things like prefetched manifest data at this level.